workflows: Rewrite build-ci-container to work on larger runners

tstellar opened the pull request:
Also switch them over to the new depot runners.

tstellar committed:
workflows: Rewrite build-ci-container to work on larger runners

Also switch them over to the new depot runners.

tstellar committed:
Remove unneeded target

tstellar committed:
Fix targets and apply new patch

tstellar committed:
Fix container name

tstellar committed:
Container push changes

tstellar committed:
Fix identation

tstellar committed:
Fix container load/save

tstellar committed:
XXX: debug

tstellar committed:
XXX: debug

tstellar committed:
XXX: debug

tstellar committed:
Revert Debug patches.

This reverts commit 5ed7ba6fc2a81f51a30c6089a31c89521a37b5dd.

This reverts commit 7985506181f09fbe54698ac811541e509070ef58.

This reverts commit 63e0fc6c3598668a7f54891ec400664d7b240414.

tstellar committed:
Fix container path name

tstellar committed:
Merge branch 'main' into container-rewrite

boomanaiden154 reviewed (commented):
Was just about to start hacking on this today. 😆 

Not too familiar with the depot runners. It seems like they're paid self hosted runners? Is the foundation paying for them?

`llvm-premerge-linux-runners` should also work and will point to self hosted runners on the GCP project handling premerge. We intend to use that runner set to run the transitioned premerge jobs once we migrate to GHA.

tstellar commented:
> Not too familiar with the depot runners. It seems like they're paid self hosted runners? Is the foundation paying for them?

The Foundation is providing a small budget for testing them out.  We're using 'Depot Managed' which means the instances are deployed in the Foundation's AWS account rather than using the machines from Depot's pool. 

> `llvm-premerge-linux-runners` should also work and will point to self hosted runners on the GCP project handling premerge. We intend to use that runner set to run the transitioned premerge jobs once we migrate to GHA.

Are these ready to use now?  I don't have a preference for which runners to use.

boomanaiden154 commented:
> The Foundation is providing a small budget for testing them out. We're using 'Depot Managed' which means the instances are deployed in the Foundation's AWS account rather than using the machines from Depot's pool.

Ah, okay. I didn't realize that had taken place. Thanks for the additional context.

> Are these ready to use now? I don't have a preference for which runners to use.

Theoretically, yes, they are ready to use and should just work. We can only run one job at a time currently though because we don't have enough quota in the region where the cluster is deployed, but that should be enough for testing. @lnihlen is working on getting additional quota. I think it would probably make more sense to use the Google hosted runners given it doesn't require any budget from the foundation, and lets us consolidate on a single system.

I think our runners will also be more powerful too (configured to be 56 vCPUs), but depends upon how the depot runners are configured.

tstellar commented:
> > The Foundation is providing a small budget for testing them out. We're using 'Depot Managed' which means the instances are deployed in the Foundation's AWS account rather than using the machines from Depot's pool.
> 
> Ah, okay. I didn't realize that had taken place. Thanks for the additional context.
> 
> > Are these ready to use now? I don't have a preference for which runners to use.
> 
> Theoretically, yes, they are ready to use and should just work. We can only run one job at a time currently though because we don't have enough quota in the region where the cluster is deployed, but that should be enough for testing. @lnihlen is working on getting additional quota. I think it would probably make more sense to use the Google hosted runners given it doesn't require any budget from the foundation, and lets us consolidate on a single system.
> 

OK, do you think I should update this PR to use those runners instead or wait a few weeks?

> I think our runners will also be more powerful too (configured to be 56 vCPUs), but depends upon how the depot runners are configured.

Depot runners are up-to 64 CPUs.  The number of CPUs is part of the runs-on label so you can choose the right size machine that fits the job.

boomanaiden154 commented:
> OK, do you think I should update this PR to use those runners instead or wait a few weeks?

I would say update it to use the new runners. This job doesn't run very often and probably shouldn't take that long to run on a decent sized machine.

> Depot runners are up-to 64 CPUs. The number of CPUs is part of the runs-on label so you can choose the right size machine that fits the job.

Ah, interesting. We don't have that much flexibility currently but could theoretically add the capability. Everything that I have seen/can imagine though either requires not much compute and should probably run on the free Github runners or requires a lot and would benefit from as many cores as possible.

tstellar committed:
Switch to using Google runners

tstellar commented:
> > OK, do you think I should update this PR to use those runners instead or wait a few weeks?
> 
> I would say update it to use the new runners. This job doesn't run very often and probably shouldn't take that long to run on a decent sized machine.
> 
> > Depot runners are up-to 64 CPUs. The number of CPUs is part of the runs-on label so you can choose the right size machine that fits the job.
> 
> Ah, interesting. We don't have that much flexibility currently but could theoretically add the capability. Everything that I have seen/can imagine though either requires not much compute and should probably run on the free Github runners or requires a lot and would benefit from as many cores as possible.

Yeah, the nice thing about the builds is that they are highly parallel, so if you are paying for runners, you may as well use the one with the most CPUs, because the cost/job ends up being roughly the same no matter how many CPUs you have.

tstellar committed:
Try using a container

tstellar committed:
Install podman

tstellar commented:
@boomanaiden154 It looks like the`llvm-premerge-linux-runners` require jobs to use a container and having nested containers doesn't work, so I may have to switch it back to the depot runners.

tstellar committed:
Revert back to depot runners

This reverts commit 1f3a36f897880591be941bd1776c5d66da04c2b1.

This reverts commit 00f4dcd0b4484bb615265eb45e9d4e642cccd285.

This reverts commit 966fd98909a301746a64076cb5678a893f43098c.

boomanaiden154 commented:
> It looks like thellvm-premerge-linux-runners require jobs to use a container and having nested containers doesn't work, so I may have to switch it back to the depot runners.

Ah, right. I forgot about that detail. My plan was to use https://github.com/GoogleContainerTools/kaniko to build things on the cluster, but that's definitely something I can get to later.

boomanaiden154 commented on .github/workflows/containers/github-action-ci/stage1.Dockerfile:
```diff
@@ -19,13 +19,11 @@ RUN curl -O -L https://github.com/llvm/llvm-project/archive/refs/tags/llvmorg-$L
 
 WORKDIR /llvm-project-llvmorg-$LLVM_VERSION
```
Can we rename this file to something other than `stage1` now that we only have a single stage?

boomanaiden154 commented on .github/workflows/containers/github-action-ci/stage1.Dockerfile:
```diff
@@ -36,7 +34,29 @@ RUN cmake -B ./build -G Ninja ./llvm \
-  -DBOOTSTRAP_CLANG_PGO_TRAINING_DATA_SOURCE_DIR=/llvm-project-llvmorg-$LLVM_VERSION/llvm
+  -DCLANG_DEFAULT_LINKER="lld"
+
+RUN ninja -C ./build stage2-clang-bolt stage2-install-distribution && ninja -C ./build install-distribution && rm -rf ./build
```
Maybe add a comment on why the `rm -rf ./build` is here? I believe it was originally to avoid out of disk space errors. Assuming the depot runners have enough disk, it would still be useful as it probably reduces checkpointing time.

tstellar replied:
I'm inclined to drop it.  I think this is a common pattern when doing something on the final image, but since this is just a builder image, I'm not sure if it makes sense to do.  I don't have a strong opinion though.

boomanaiden154 replied:
Seems reasonable enough to me.

boomanaiden154 commented on .github/workflows/build-ci-container.yml:
```diff
@@ -18,95 +18,70 @@ on:
-    runs-on: ubuntu-latest
+    runs-on: depot-ubuntu-24.04-16
+    outputs:
+      container-name: ${{ steps.vars.outputs.container-name }}
```
Why do we need to set these as outputs?

tstellar replied:
They're used in the push-ci-container job.

boomanaiden154 commented on .github/workflows/build-ci-container.yml:
```diff
@@ -18,95 +18,70 @@ on:
+      - name: Save container image
         run: |
-          podman load -i /mnt/podman/stage1-toolchain.tar
+          podman save  ${{ steps.vars.outputs.container-name-tag }} >  ${{ steps.vars.outputs.container-filename }}
```
Maybe add a comment here on why we're saving it? Wasn't immediately obvious to me before reading below.

tstellar committed:
Review comments

boomanaiden154 reviewed (approved):
LGTM after addressing the final comment.

tstellar committed:
Remove redundant -f Dockerfile

tstellar committed:
Switch to using older ubuntu version

We can save money by only supporting on ubuntu version in depot and we need to use the oldest version when building the release binaries.

boomanaiden154 commented:
@tstellar What are your plans on landing this? I wanted to try porting it to the premerge cluster this week, but would prefer to build off of this patch in tree.
